File: 0785-is-graph-bipartite/0785-is-graph-bipartite.py
class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        color = [-1]*len(graph)
        for i in range(0, len(graph)):
            if color[i] == -1:
                if not self.dfs(i, graph, color): return False
        return True
    def dfs(self, src, graph, color):
        q = deque()
        q.append(src)
        color[src] = 1
        while q:
            node = q.popleft()
            for x in graph[node]:
                if color[x] == -1:
                    color[x] = 1 - color[node]
                    q.append(x)
                elif color[x] == color[node]: return False
        return True

File: 0785-is-graph-bipartite/0785-is-graph-bipartite.java
class Solution {
    public boolean isBipartite(int[][] graph) {
        int[] color = new int[graph.length];
        Arrays.fill(color, -1);
        for(int i=0; i<graph.length; i++){
            if(color[i] == -1){
                if(!dfs(i, graph, color)) return false;
            }
        }
        return true;
    }
    boolean dfs(int src, int[][] graph, int[] color){
        Queue<Integer> q = new LinkedList<>();
        q.add(src);
        color[src] = 1;
        while(!q.isEmpty()){
            int node = q.poll();
            for(int x : graph[node]){
                if(color[x] == -1){
                    color[x] = 1 - color[node];
                    q.add(x);
                }
                else if(color[x] == color[node]) return false;
            }
        }
        return true;
    }
}